File: swift/examples/activity_feed/activity_feed.swift
import Foundation
import Uimd

private let kDefaultActivityType = "Info"
private let kTimestampEnv = "UIMD_ACTIVITY_FEED_TIMESTAMP"
private let kInitialSampleCount = 2
private let kHeaderLayoutTestWidth = 112
private let kHeaderLayoutTestHeight = 38

private struct SampleActivity
{
    let type: String
    let message: String
}

private let kSampleActivities: [SampleActivity] = [
    SampleActivity(type: "Info", message: "Workspace opened"),
    SampleActivity(type: "Task", message: "Report export queued"),
    SampleActivity(type: "Deploy", message: "Version 1.4 deployed"),
    SampleActivity(type: "Warning", message: "Review queue is almost full"),
    SampleActivity(type: "Note", message: "Follow-up scheduled"),
    SampleActivity(type: "Task", message: "Background sync finished"),
]

private struct ActivityItemData
{
    var timestamp: String
    var eventType: String
    var message: String
    var showTimestamp = true
}

private struct ConstrainedPadding
{
    var top = 0
    var right = 0
    var bottom = 0
    var left = 0
}

private func trim(_ text: String) -> String
{
    text.trimmingCharacters(in: .whitespacesAndNewlines)
}

private func currentTimestamp() -> String
{
    if let fixed = ProcessInfo.processInfo.environment[kTimestampEnv], !fixed.isEmpty
    {
        return fixed
    }
    let formatter = DateFormatter()
    formatter.dateFormat = "HH:mm"
    return formatter.string(from: Date())
}

private func selectedComboValue(_ comboBox: ComboBox, fallback: String) -> String
{
    guard comboBox.selectedIndex >= 0 && comboBox.selectedIndex < comboBox.options.count else
    {
        return fallback
    }
    return comboBox.options[comboBox.selectedIndex]
}

private func selectComboValue(_ comboBox: ComboBox, _ value: String)
{
    if let index = comboBox.options.firstIndex(of: value)
    {
        comboBox.selectedIndex = index
    }
    else
    {
        comboBox.selectedIndex = comboBox.options.isEmpty ? -1 : 0
    }
}

private func renderedText(_ rendered: [[TerminalCell]]) -> String
{
    rendered.map { row in row.map(\.text).joined() }.joined(separator: "\n")
}

private func splitLines(_ text: String) -> [String]
{
    let parts = text.split(separator: "\n", omittingEmptySubsequences: false).map(String.init)
    return parts.isEmpty ? [""] : parts
}

private func wrappedLineCount(_ text: String, width: Int) -> Int
{
    let safeWidth = max(1, width)
    return max(1, splitLines(text).reduce(0) { count, line in
        count + max(1, (line.count + safeWidth - 1) / safeWidth)
    })
}

private func generatedSourceHeight(_ view: GeneratedWindowBase) -> Int
{
    max(1, view.layout.reduce(0) { height, entry in
        max(height, entry.sourceCell.row + entry.sourceCell.height)
    })
}

private func compactContentHeight(_ view: GeneratedWindowBase, messageRows: Int) -> Int
{
    max(1, view.layout.reduce(1) { height, entry in
        if entry.name.isEmpty
        {
            return height
        }
        let entryHeight = entry.name == "message"
            ? max(1, messageRows)
            : max(1, entry.charsSize.height)
        return max(height, entry.relative.row + entryHeight)
    })
}

private func markerColor(_ eventType: String) -> Color
{
    switch eventType
    {
    case "Task":
        return Color("#a78bfa")
    case "Warning":
        return Color("#facc15")
    case "Deploy":
        return Color("#22c55e")
    case "Note":
        return Color("#fb7185")
    default:
        return Color("#38bdf8")
    }
}

private func applyLeftMarker(_ rendered: inout [[TerminalCell]], color: Color)
{
    for row in rendered.indices
    {
        guard !rendered[row].isEmpty else
        {
            continue
        }
        rendered[row][0].text = "▐"
        rendered[row][0].foreground = color
    }
}

private func renderActivityItem(_ item: ActivityItemData, width: Int) -> [[TerminalCell]]
{
    let view = ActivityItemUI()
    view.timestamp.setText(item.showTimestamp ? item.timestamp : "")
    view.event_type.setText(item.eventType)
    view.message.setText(item.message)

    _ = view.renderContent(
        size: Size(width: max(1, width), height: generatedSourceHeight(view)),
        focusedName: nil,
        editMode: false
    )
    let messageRows = wrappedLineCount(item.message, width: max(1, view.message.frame.width))
    var rendered = view.renderContent(
        size: Size(width: max(1, width), height: compactContentHeight(view, messageRows: messageRows)),
        focusedName: nil,
        editMode: false
    )
    applyLeftMarker(&rendered, color: markerColor(item.eventType))
    return rendered
}

private final class ActivityFeedPanel: ActivityFeedPanelUI
{
    private var activities: [ActivityItemData] = []
    private var autoScrollEnabled = false

    override init()
    {
        super.init()
        setDynamicChildrenRenderer
        {
            [weak self] width in
            self?.renderActivities(width: width) ?? []
        }
    }

    func clearActivities()
    {
        activities.removeAll()
        clearChildren()
        invalidateDynamicChildren()
    }

    func appendActivity(timestamp: String, eventType: String, message: String, showTimestamp: Bool)
    {
        activities.append(ActivityItemData(
            timestamp: timestamp,
            eventType: eventType,
            message: message,
            showTimestamp: showTimestamp
        ))
        invalidateDynamicChildren()
        if autoScrollEnabled
        {
            scrollToBottom()
        }
    }

    override func setAutoScroll(_ enabled: Bool)
    {
        autoScrollEnabled = enabled
        scrollView().setAutoScroll(enabled)
        if autoScrollEnabled
        {
            scrollToBottom()
        }
    }

    func activityCount() -> Int
    {
        activities.count
    }

    private func renderActivities(width: Int) -> [[[TerminalCell]]]
    {
        activities.map { renderActivityItem($0, width: max(1, width)) }
    }
}

private struct SettingsResult
{
    var autoScroll = true
    var showTimestamps = true
    var defaultType = kDefaultActivityType
}

private final class SettingsDialog: SettingsUI
{
    var onCancel: (() -> Void)?
    private(set) var isOpen = false

    func configure(_ settings: SettingsResult)
    {
        auto_scroll.setChecked(settings.autoScroll)
        show_timestamps.setChecked(settings.showTimestamps)
        selectDefaultType(settings.defaultType)
    }

    func open()
    {
        isOpen = true
    }

    func close()
    {
        isOpen = false
    }

    func result() -> SettingsResult
    {
        SettingsResult(
            autoScroll: auto_scroll.checked,
            showTimestamps: show_timestamps.checked,
            defaultType: selectedDefaultType()
        )
    }

    func selectDefaultType(_ value: String)
    {
        selectComboValue(default_type, value)
    }

    func selectedDefaultType() -> String
    {
        selectedComboValue(default_type, fallback: kDefaultActivityType)
    }

    override func onPreviewKey(_ event: KeyEvent) -> Bool
    {
        if event.key == "Escape" && !event.editMode
        {
            onCancel?()
            return true
        }
        return false
    }
}

private final class ActivityFeedApp: ActivityFeedUI
{
    private let panel = ActivityFeedPanel()
    private let settingsDialog = SettingsDialog()
    private let modalStack = GeneratedWindowStack()
    private var autoScroll = true
    private var showTimestamps = true
    private var quitRequested = false
    private var defaultType = kDefaultActivityType

    override init()
    {
        super.init()
        feed.setChild(panel)
    }

    func open()
    {
        selectComboValue(activity_type, defaultType)
        _ = renderContent(size: generatedContentSize(), focusedName: nil, editMode: false)
        panel.setAutoScroll(autoScroll)
        if panel.activityCount() == 0
        {
            seedInitial()
        }
    }

    func button(_ name: String)
    {
        if name == "add_button"
        {
            addCurrentActivity()
        }
        else if name == "seed_button"
        {
            seedActivities()
        }
        else if name == "clear_button"
        {
            clearActivities()
        }
        else if name == "settings_button"
        {
            openSettings()
        }
        else if name == "quit_button"
        {
            quitRequested = true
        }
    }

    func addCurrentActivity()
    {
        let activityMessage = trim(message.value)
        if activityMessage.isEmpty
        {
            status.setText("Enter an activity message.")
            return
        }
        appendActivity(eventType: selectedComboValue(activity_type, fallback: defaultType), messageText: activityMessage)
        message.setValue("")
        message.cursor = 0
        updateCountStatus()
    }

    func appendActivity(eventType: String, messageText: String)
    {
        panel.appendActivity(
            timestamp: currentTimestamp(),
            eventType: eventType,
            message: messageText,
            showTimestamp: showTimestamps
        )
    }

    func seedActivities()
    {
        for activity in kSampleActivities
        {
            appendActivity(eventType: activity.type, messageText: activity.message)
        }
        updateCountStatus()
    }

    func clearActivities()
    {
        panel.clearActivities()
        status.setText("Feed cleared")
    }

    func activityCount() -> Int
    {
        panel.activityCount()
    }

    func hasModalOverlay() -> Bool
    {
        settingsDialog.isOpen
    }

    override func runtimeOptions() -> GeneratedWindowRuntimeOptions
    {
        var options = super.runtimeOptions()
        options.windowStack = modalStack
        return options
    }

    override func onAddButtonClick()
    {
        addCurrentActivity()
    }

    override func onSeedButtonClick()
    {
        seedActivities()
    }

    override func onClearButtonClick()
    {
        clearActivities()
    }

    override func onSettingsButtonClick()
    {
        openSettings()
    }

    override func onQuitButtonClick()
    {
        quitRequested = true
    }

    override func onMessageSubmit(_ value: String)
    {
        message.setValue(value)
    }

    override func shouldClose() -> Bool
    {
        quitRequested
    }

    private func openSettings()
    {
        settingsDialog.configure(SettingsResult(
            autoScroll: autoScroll,
            showTimestamps: showTimestamps,
            defaultType: defaultType
        ))
        settingsDialog.open()
        settingsDialog.onCancel = { [weak self] in
            guard let self else
            {
                return
            }
            status.setText("Settings canceled")
            closeSettings()
        }
        var frame = GeneratedWindowFrameOptions()
        frame.onButton = { [weak self] name in
            guard let self else
            {
                return false
            }
            if name == "save_button"
            {
                applySettings(settingsDialog.result())
            }
            else
            {
                status.setText("Settings canceled")
            }
            closeSettings()
            return true
        }
        modalStack.push(settingsDialog, frame)
    }

    private func closeSettings()
    {
        modalStack.remove(settingsDialog)
        settingsDialog.close()
    }

    private func applySettings(_ settings: SettingsResult)
    {
        autoScroll = settings.autoScroll
        showTimestamps = settings.showTimestamps
        defaultType = settings.defaultType.isEmpty ? kDefaultActivityType : settings.defaultType
        selectComboValue(activity_type, defaultType)
        panel.setAutoScroll(autoScroll)
        status.setText("Settings saved")
    }

    private func seedInitial()
    {
        let count = min(kInitialSampleCount, kSampleActivities.count)
        for index in 0..<count
        {
            appendActivity(eventType: kSampleActivities[index].type, messageText: kSampleActivities[index].message)
        }
        updateCountStatus()
    }

    private func updateCountStatus()
    {
        status.setText("\(panel.activityCount()) activities")
    }
}

private func smoke()
{
    let window = ActivityFeedUI()
    precondition(window.windowTitle == "Activity Feed")
    precondition(window.header_title != nil)
    precondition(window.settings_button != nil)
    precondition(window.quit_button != nil)
    precondition(window.feed != nil)
    precondition(window.activity_type != nil)
    precondition(window.message != nil)
    precondition(window.add_button != nil)
    precondition(window.seed_button != nil)
    precondition(window.clear_button != nil)
    precondition(window.status != nil)

    _ = window.renderContent(
        size: Size(width: kHeaderLayoutTestWidth, height: kHeaderLayoutTestHeight),
        focusedName: nil,
        editMode: false
    )
    precondition(window.header_title.frame.col + window.header_title.frame.width <= window.settings_button.frame.col)
    precondition(window.settings_button.frame.col + window.settings_button.frame.width < window.quit_button.frame.col)
}

private func logicTest()
{
    let app = ActivityFeedApp()
    app.open()

    precondition(app.activityCount() == kInitialSampleCount)
    precondition(app.feed.child != nil)
    let feedSize = Size(width: max(1, app.feed.frame.width), height: max(1, app.feed.frame.height))
    var feedText = renderedText(app.feed.render(size: feedSize))
    precondition(feedText.contains("Workspace opened"))

    app.message.setValue("Manual event")
    app.button("add_button")
    precondition(app.message.value.isEmpty)
    precondition(app.status.text == "3 activities")
    feedText = renderedText(app.feed.render(size: feedSize))
    precondition(feedText.contains("Manual event"))

    app.seedActivities()
    precondition(app.activityCount() == kInitialSampleCount + 1 + kSampleActivities.count)
    feedText = renderedText(app.feed.render(size: feedSize))
    precondition(feedText.contains("Background sync finished"))

    app.clearActivities()
    precondition(app.activityCount() == 0)
    precondition(app.status.text == "Feed cleared")

    app.button("settings_button")
    precondition(app.hasModalOverlay())
    app.button("quit_button")
    let options = app.runtimeOptions()
    precondition(options.shouldClose?() == true)
}

@main
enum Program
{
    static func main()
    {
        let code = GeneratedWindowRuntime.runGeneratedAppMain
        {
            if CommandLine.arguments.contains("--smoke")
            {
                smoke()
                return 0
            }
            if CommandLine.arguments.contains("--logic-test")
            {
                logicTest()
                return 0
            }
            let app = ActivityFeedApp()
            app.open()
            return GeneratedWindowRuntime.runGeneratedWindow(app, app.runtimeOptions(), CommandLine.arguments)
        }
        Foundation.exit(Int32(code))
    }
}

File: go/examples/activity_feed/activity_feed.go
package main

import (
	"fmt"
	"os"
	"strings"
	"time"

	activityfeedpanel "uimd_examples/activity_feed/activity_feed_panel"
	settingsdialog "uimd_examples/activity_feed/settings"

	uimd "uimd"
)

const (
	defaultActivityType = "Info"
	timestampEnv        = "UIMD_ACTIVITY_FEED_TIMESTAMP"
)

type sampleActivity struct {
	eventType string
	message   string
}

var sampleActivities = []sampleActivity{
	{"Info", "Workspace opened"},
	{"Task", "Report export queued"},
	{"Deploy", "Version 1.4 deployed"},
	{"Warning", "Review queue is almost full"},
	{"Note", "Follow-up scheduled"},
	{"Task", "Background sync finished"},
}

const initialSampleCount = 2

type ActivityFeedApp struct {
	*ActivityFeedUI
	feed             *activityfeedpanel.ActivityFeedPanel
	settingsDialog   *settingsdialog.SettingsDialog
	stack            uimd.GeneratedWindowStack
	autoScroll       bool
	showTimestamps   bool
	quitRequested    bool
	defaultType      string
	settingsWasShown bool
}

func NewActivityFeedApp() *ActivityFeedApp {
	panel := activityfeedpanel.NewActivityFeedPanel()
	app := &ActivityFeedApp{
		ActivityFeedUI: NewActivityFeedUI(),
		feed:           panel,
		settingsDialog: settingsdialog.NewSettingsDialog(),
		autoScroll:     true,
		showTimestamps: true,
		defaultType:    defaultActivityType,
	}
	app.SetEventHandler(app)
	app.Feed.SetChild(panel)
	return app
}

func (app *ActivityFeedApp) Open() {
	selectComboValue(app.ActivityType, app.defaultType)
	uimd.RenderGeneratedWindowContent(app.GeneratedWindowBase, uimd.Size{Width: 100, Height: 32}, -1, nil, false)
	app.feed.SetAutoScroll(app.autoScroll)
	if app.feed.ActivityCount() == 0 {
		app.seedInitial()
	}
}

func (app *ActivityFeedApp) RuntimeOptions() uimd.GeneratedWindowRuntimeOptions {
	options := app.ActivityFeedUI.RuntimeOptions()
	options.WindowStack = &app.stack
	options.ShouldClose = app.ShouldClose
	return options
}

func (app *ActivityFeedApp) ShouldClose() bool {
	return app.quitRequested || app.GeneratedWindowBase.ShouldClose()
}

func (app *ActivityFeedApp) OnAddButtonClick() {
	app.AddCurrentActivity()
}

func (app *ActivityFeedApp) OnSeedButtonClick() {
	app.SeedActivities()
}

func (app *ActivityFeedApp) OnClearButtonClick() {
	app.ClearActivities()
}

func (app *ActivityFeedApp) OnSettingsButtonClick() {
	app.OpenSettings()
}

func (app *ActivityFeedApp) OnQuitButtonClick() {
	app.quitRequested = true
	app.RequestClose()
}

func (app *ActivityFeedApp) OnMessageSubmit(value string) {
	app.Message.SetValue(value)
}

func (app *ActivityFeedApp) AddCurrentActivity() {
	message := strings.TrimSpace(app.Message.Value)
	if message == "" {
		app.Status.SetText("Enter an activity message.")
		return
	}
	app.AppendActivity(selectedComboValue(app.ActivityType, app.defaultType), message)
	app.Message.SetValue("")
	app.Message.SetCursor(0)
	app.updateCountStatus()
}

func (app *ActivityFeedApp) AppendActivity(eventType string, message string) {
	app.feed.AppendActivity(currentTimestamp(), eventType, message, app.showTimestamps)
}

func (app *ActivityFeedApp) SeedActivities() {
	for _, activity := range sampleActivities {
		app.AppendActivity(activity.eventType, activity.message)
	}
	app.updateCountStatus()
}

func (app *ActivityFeedApp) ClearActivities() {
	app.feed.ClearActivities()
	app.Status.SetText("Feed cleared")
}

func (app *ActivityFeedApp) ActivityCount() int {
	return app.feed.ActivityCount()
}

func (app *ActivityFeedApp) HasModalOverlay() bool {
	return !app.stack.Empty()
}

func (app *ActivityFeedApp) OpenSettings() {
	app.settingsDialog.Configure(settingsdialog.SettingsResult{
		AutoScroll:     app.autoScroll,
		ShowTimestamps: app.showTimestamps,
		DefaultType:    app.defaultType,
	})
	app.settingsDialog.Open()
	app.settingsDialog.OnCancel = func() {
		app.Status.SetText("Settings canceled")
		app.closeSettings()
	}
	frame := uimd.GeneratedWindowFrameOptions{
		ClassName: "Settings",
		OnButton: func(name string) {
			if name == "save_button" {
				app.applySettings(app.settingsDialog.Result())
			} else {
				app.Status.SetText("Settings canceled")
			}
			app.closeSettings()
		},
	}
	app.settingsWasShown = true
	app.stack.Push(app.settingsDialog, frame)
}

func (app *ActivityFeedApp) closeSettings() {
	app.stack.Remove(app.settingsDialog.UI)
	app.settingsDialog.Close()
}

func (app *ActivityFeedApp) applySettings(result settingsdialog.SettingsResult) {
	app.autoScroll = result.AutoScroll
	app.showTimestamps = result.ShowTimestamps
	app.defaultType = result.DefaultType
	if app.defaultType == "" {
		app.defaultType = defaultActivityType
	}
	selectComboValue(app.ActivityType, app.defaultType)
	app.feed.SetAutoScroll(app.autoScroll)
	app.Status.SetText("Settings saved")
}

func (app *ActivityFeedApp) seedInitial() {
	count := initialSampleCount
	if count > len(sampleActivities) {
		count = len(sampleActivities)
	}
	for index := 0; index < count; index++ {
		app.AppendActivity(sampleActivities[index].eventType, sampleActivities[index].message)
	}
	app.updateCountStatus()
}

func (app *ActivityFeedApp) updateCountStatus() {
	app.Status.SetText(fmt.Sprintf("%d activities", app.feed.ActivityCount()))
}

func currentTimestamp() string {
	if fixed := os.Getenv(timestampEnv); fixed != "" {
		return fixed
	}
	return time.Now().Format("15:04")
}

func selectedComboValue(combo *uimd.ComboBox, fallback string) string {
	index := combo.SelectedIndex
	if index < 0 || index >= len(combo.Options) {
		return fallback
	}
	return combo.Options[index]
}

func selectComboValue(combo *uimd.ComboBox, value string) {
	for index, option := range combo.Options {
		if option == value {
			combo.SetSelectedIndex(index)
			return
		}
	}
	combo.SetSelectedIndex(0)
}

func smoke() {
	window := NewActivityFeedUI()
	if window.Title() != "Activity Feed" || window.HeaderTitle == nil || window.SettingsButton == nil ||
		window.QuitButton == nil || window.Feed == nil || window.ActivityType == nil ||
		window.Message == nil || window.AddButton == nil || window.SeedButton == nil ||
		window.ClearButton == nil || window.Status == nil {
		panic("activity feed smoke failed")
	}
}

func logicTest() {
	app := NewActivityFeedApp()
	app.Open()
	if app.ActivityCount() != initialSampleCount {
		panic("initial seed failed")
	}
	app.Message.SetValue("Manual event")
	app.AddCurrentActivity()
	if app.Message.Value != "" || app.Status.Text != "3 activities" {
		panic("manual activity failed")
	}
	app.SeedActivities()
	if app.ActivityCount() != initialSampleCount+1+len(sampleActivities) {
		panic("seed activities failed")
	}
	app.ClearActivities()
	if app.ActivityCount() != 0 || app.Status.Text != "Feed cleared" {
		panic("clear failed")
	}
	app.OpenSettings()
	if !app.HasModalOverlay() {
		panic("settings modal not open")
	}
	app.OnQuitButtonClick()
	if !app.ShouldClose() {
		panic("quit did not request close")
	}
}

func main() {
	code := uimd.RunGeneratedAppMain(func() int {
		if len(os.Args) > 1 {
			switch os.Args[1] {
			case "--smoke":
				smoke()
				return 0
			case "--logic-test":
				logicTest()
				return 0
			}
		}
		app := NewActivityFeedApp()
		app.Open()
		return uimd.RunGeneratedWindow(app, app.RuntimeOptions(), os.Args)
	})
	os.Exit(code)
}
